samples/nrf9160/modem_shell/src/utils/mosh_print.c

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <assert.h>
+
+#include <zephyr.h>
+#include <posix/time.h>
+#include <shell/shell.h>
+#include <shell/shell_uart.h>
+
+#include "mosh_print.h"
+
+extern const struct shell *shell_global;
+
+/** Configuration on whether timestamps are added to the print. */
+bool mosh_print_timestamp_use;
+
+/** Buffer used for printing the timestamp. */
+static char timestamp_str[30];
+/** Buffer used for generating the text to be printed. */
+static char mosh_print_buf[256];
+/** Mutex for protecting mosh_print_buf */
+K_MUTEX_DEFINE(mosh_print_buf_mutex);
+
+static const char *create_timestamp_string(void)
+{
+	uint32_t year;
+	uint32_t month;
+	uint32_t day;
+
+	uint32_t hours;
+	uint32_t mins;
+	uint32_t secs;
+	uint32_t msec;
+
+	struct timespec tp;
+	struct tm ltm = { 0 };
+
+	clock_gettime(CLOCK_REALTIME, &tp);
+	gmtime_r(&tp.tv_sec, &ltm);
+
+	msec = tp.tv_nsec / 1000000;
+	secs = ltm.tm_sec;
+	mins = ltm.tm_min;
+	hours = ltm.tm_hour;
+	day = ltm.tm_mday;
+	month = ltm.tm_mon;
+	year = 1900 + ltm.tm_year;
+
+	sprintf(timestamp_str,
+		"[%04d-%02d-%02d %02d:%02d:%02d.%03d] ",
+		year, month, day, hours, mins, secs, msec);
+
+	return timestamp_str;
+}
+
+void mosh_fprintf(enum mosh_print_level print_level, const char *fmt, ...)
+{
+	va_list args;
+	int chars = 0;
+	const char *timestamp;
+
+	k_mutex_lock(&mosh_print_buf_mutex, K_FOREVER);
+
+	if (shell_global == NULL) {
+		printk("CANNOT PRINT BECAUSE NO SHELL BACKEND!!!");
+		shell_global = shell_backend_uart_get_ptr();
+		if (shell_global == NULL) {
+			printk("STILL CANNOT PRINT BECAUSE CANNOT OBTAIN SHELL BACKEND!!!");
+			vsnprintf(mosh_print_buf, sizeof(mosh_print_buf), fmt, args);
+			printk("%s", mosh_print_buf);
+			return;
+		}
+	}
+
+	va_start(args, fmt);
+
+	/* Add timestamp to print buffer if requested */
+	if (mosh_print_timestamp_use) {
+		timestamp = create_timestamp_string();
+		chars = snprintf(mosh_print_buf, sizeof(mosh_print_buf), "%s", timestamp);
+		if (chars < 0) {
+			shell_error(shell_global, "Error while printing timestamp...");
+			chars = 0;
+		}
+	}
+
+	/* Add requested printf-like string */
+	chars += vsnprintf(mosh_print_buf + chars, sizeof(mosh_print_buf) - chars, fmt, args);
+	if (chars >= sizeof(mosh_print_buf) - 1) {
```

Comment:
Check if the -1 in needed here? I think not.

Reply:
True. This was some leftover from testing how NUL termination is handled but snprintf will include NUL termination to the given buffer length and return value is how many characters would have been written if there is space. Hence just removing -1 is accurate.